File: server/src/routes/flows.ts
import { Router } from 'express';
import { flowService } from '../services/flow.js';

const router = Router();

// Create a new flow
router.post('/', async (req, res) => {
  try {
    const { name, description, initialState, states } = req.body;

    if (!name || !initialState || !states) {
      return res.status(400).json({ error: 'name, initialState, and states are required' });
    }

    const flow = await flowService.createFlow({
      name,
      description,
      initialState,
      states,
    });

    res.status(201).json(flow);
  } catch (error) {
    console.error('Error creating flow:', error);
    const message = error instanceof Error ? error.message : 'Failed to create flow';
    res.status(500).json({ error: message });
  }
});

// List all flows
router.get('/', async (req, res) => {
  try {
    const flowsList = await flowService.listFlows();
    res.json(flowsList);
  } catch (error) {
    console.error('Error listing flows:', error);
    res.status(500).json({ error: 'Failed to list flows' });
  }
});

// Get a specific flow
router.get('/:id', async (req, res) => {
  try {
    const flow = await flowService.getFlow(parseInt(req.params.id));
    if (!flow) {
      return res.status(404).json({ error: 'Flow not found' });
    }
    res.json(flow);
  } catch (error) {
    console.error('Error getting flow:', error);
    res.status(500).json({ error: 'Failed to get flow' });
  }
});

// Start a new flow execution (trigger)
router.post('/:id/execute', async (req, res) => {
  try {
    const flowId = parseInt(req.params.id);
    const { context } = req.body;

    const execution = await flowService.startExecution(flowId, context || {});
    res.status(201).json(execution);
  } catch (error) {
    console.error('Error starting execution:', error);
    const message = error instanceof Error ? error.message : 'Failed to start execution';
    res.status(500).json({ error: message });
  }
});

// Get execution status
router.get('/executions/:executionId', async (req, res) => {
  try {
    const execution = await flowService.getExecution(parseInt(req.params.executionId));
    if (!execution) {
      return res.status(404).json({ error: 'Execution not found' });
    }
    res.json(execution);
  } catch (error) {
    console.error('Error getting execution:', error);
    res.status(500).json({ error: 'Failed to get execution' });
  }
});

// Send an event to a paused execution
router.post('/executions/:executionId/event', async (req, res) => {
  try {
    const executionId = parseInt(req.params.executionId);
    const { event, data } = req.body;

    if (!event) {
      return res.status(400).json({ error: 'event is required' });
    }

    const execution = await flowService.sendEvent(executionId, event, data || {});
    res.json(execution);
  } catch (error) {
    console.error('Error sending event:', error);
    const message = error instanceof Error ? error.message : 'Failed to send event';
    res.status(500).json({ error: message });
  }
});

export default router;

File: server/app/api/api_v1/endpoints/flows.py
from typing import List, Any, Optional, Dict
from fastapi import APIRouter, Depends, HTTPException, Response
from sqlmodel import Session
from app.core.db import get_session
from app.models.flow import Flow, FlowExecution
from app.services.flow_service import flow_service

router = APIRouter()

@router.get("", response_model=List[Flow])
def list_flows(
    session: Session = Depends(get_session)
) -> Any:
    return flow_service.list_flows(session)

@router.post("", response_model=Flow)
def create_flow(
    data: Dict[str, Any],
    session: Session = Depends(get_session)
) -> Any:
    return flow_service.create_flow(session, data)

@router.get("/{id}", response_model=Flow)
def get_flow(
    id: int,
    session: Session = Depends(get_session)
) -> Any:
    flow = flow_service.get_flow(session, id)
    if not flow:
        raise HTTPException(status_code=404, detail="Flow not found")
    return flow

@router.patch("/{id}", response_model=Flow)
def update_flow(
    id: int,
    data: Dict[str, Any],
    session: Session = Depends(get_session)
) -> Any:
    flow = flow_service.update_flow(session, id, data)
    if not flow:
        raise HTTPException(status_code=404, detail="Flow not found")
    return flow

@router.delete("/{id}")
def delete_flow(
    id: int,
    session: Session = Depends(get_session)
) -> Any:
    if not flow_service.delete_flow(session, id):
        raise HTTPException(status_code=404, detail="Flow not found")
    return Response(status_code=204)

@router.post("/{id}/execute", response_model=FlowExecution)
def execute_flow(
    id: int,
    data: Dict[str, Any],
    session: Session = Depends(get_session)
) -> Any:
    return flow_service.start_execution(session, id, data.get("context", {}))

@router.get("/executions/{id}", response_model=FlowExecution)
def get_execution(
    id: int,
    session: Session = Depends(get_session)
) -> Any:
    execution = flow_service.get_execution(session, id)
    if not execution:
        raise HTTPException(status_code=404, detail="Execution not found")
    return execution
